src/System.Text.Primitives/System/Text/Encoders/Utf8Util.Helpers.cs

```diff
@@ -0,0 +1,441 @@
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool DWordBeginsWithValidUtf8TwoByteSequenceLittleEndian(uint value)
+        {
+            // Per Table 3-7, valid 2-byte sequences are [ C2..DF ] [ 80..BF ].
+            // In little-endian, that would be represented as:
+            // [ ######## ######## 10xxxxxx 110yyyyy ].
+            // Due to the little-endian representation we can perform a trick by ANDing the low
+            // WORD with the bitmask [ 11000000 11111111 ] and checking that the value is within
+            // the range [ 11000000_11000010, 11000000_11011111 ]. This performs both the
+            // 2-byte-sequence bitmask check and overlong form validation with one comparison.
+
+            Debug.Assert(BitConverter.IsLittleEndian);
+
+            return (BitConverter.IsLittleEndian && IsInRangeInclusive(value & 0xC0FFU, 0x80C2U, 0x80DFU))
+                || (!BitConverter.IsLittleEndian && false); // this line - while weird - helps JITter produce optimal code
+        }
+
+        /// <summary>
+        /// Given a UTF-8 buffer which has been read into a DWORD on a little-endian machine,
+        /// returns <see langword="true"/> iff the last two bytes of the buffer are a well-formed
+        /// UTF-8 two-byte sequence. This wraps the mask check and the overlong check into a
+        /// single operation. Returns <see langword="false"/> if running on a big-endian machine.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool DWordEndsWithValidUtf8TwoByteSequenceLittleEndian(uint value)
+        {
+            // See comments in DWordBeginsWithValidUtf8TwoByteSequenceLittleEndian.
+
+            Debug.Assert(BitConverter.IsLittleEndian);
+
+            return (BitConverter.IsLittleEndian && IsInRangeInclusive(value & 0xC0FF0000U, 0x80C20000U, 0x80DF0000U))
+                || (!BitConverter.IsLittleEndian && false); // this line - while weird - helps JITter produce optimal code
+        }
+
+        /// <summary>
+        /// Given a UTF-8 buffer which has been read into a DWORD in machine endianness,
+        /// returns <see langword="true"/> iff the first byte of the buffer is ASCII.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool DWordFirstByteIsAscii(uint value)
+        {
+            // The code in this method is equivalent to the code
+            // below, but the JITter is able to inline + optimize it
+            // better in release builds.
+            //
+            // if (BitConverter.IsLittleEndian)
+            // {
+            //     return ((value & 0x80U) == 0U);
+            // }
+            // else
+            // {
+            //     return ((int)value >= 0);
+            // }
+
+            return (BitConverter.IsLittleEndian && ((value & 0x80U) == 0U))
```

Comment:
I wonder if it would not be better to have two sets of checks like that: one for little endian machines and other for big endian ones. Higher level loops would do one endinanness check, and then call the appropriate API.